FullStorySegmentMiddleware/Classes/FullStoryMiddleware.m

```diff
@@ -125,102 +127,148 @@ - (SEGPayload *) getNewPayloadWithFSURL:(SEGContext * _Nonnull)context {
-            [dict setObject:tempDict forKey:[key stringByReplacingOccurrencesOfString:@"_" withString:@""]];
-        }else if([item isKindOfClass:[NSArray class]]){
+        
+        if ([item isKindOfClass:[NSDictionary class]]) {
+            // if array of dicts, we then loop through all dicts, flatten out each key/val into arrays, we will loose the object association but it allows user to search for each key/val in the array in FS (i.e. searching for one product when array of products are sent)
+//
+            NSMutableDictionary *tempDict = [[NSMutableDictionary alloc] initWithDictionary:[self getSuffixedProps:(NSDictionary *) item]];
+            NSLog(@"tempDict: %@, key %@",tempDict, key);
+//            [tempDict addEntriesFromDictionary:dict[key]];
+//            [dict setObject:tempDict forKey:[key stringByReplacingOccurrencesOfString:@"_" withString:@""]];
+            [self appendToDictionary:dict withKey:key andDictionary:tempDict];
+            NSLog(@"dict: %@",dict);
+        } else if ([item isKindOfClass:[NSArray class]]) {
             // TODO: Segment spec should not allow nested array properties, ignore for now, but we should handle it eventually
-            NSMutableDictionary *tempDict = [[NSMutableDictionary alloc]initWithDictionary:@{[@"idx" stringByAppendingFormat:@"%d",i]:item}];
-            [tempDict addEntriesFromDictionary:[dict valueForKey:key]];
-            [dict setObject:tempDict forKey:[key stringByReplacingOccurrencesOfString:@"_" withString:@""]];
-        }else{
-            //default to simple object
-            NSString* suffix = [[self getSuffixStringFromSimpleObject:item] stringByAppendingString:@"s"];
+            NSDictionary *tempDict = [self getDictionaryFromArrayObject:(NSArray *)item withKey:key];
+            [self appendToDictionary:dict withKey:key andDictionary:tempDict];
+        } else {
+            // default to simple object
             // if there are arrays of mixed type, then in the final props we will add approporate values to the same, key but with each type suffix
-            // get the current array form this suffix, if any, then append current item
+            NSString* suffix = [[self getSuffixStringFromSimpleObject:item] stringByAppendingString:@"s"];
+            // get the current array form this specified suffix, if any, then append current item
             NSMutableArray *tempArr = [[NSMutableArray alloc] initWithArray:[dict valueForKey:[key stringByAppendingString:suffix]]];
             [tempArr addObject:item];
             [dict setValue:tempArr forKey:[key stringByAppendingString:suffix]];
         }
     }
+
     return dict;
 }
 
+-(void) appendToDictionary:(NSMutableDictionary *) dict withKey:(NSString *) key andDictionary:(NSDictionary *) dict2 {
+    // when adding a parsed array into the result dict, check if the key with suffix already exsist, if so then we need to append to the result arrays insead of replacing the object.
```

Comment:
nit: wrap long comment into multiple lines (apply this throughout the entire file)